### PR TITLE
Fix for #851 and #852

### PR DESCRIPTION
**Pull request by BlueRiverInteractive**
This is the process:
1. AppCompiler walks through resource folder in search of `.storyboard` and `.xib` files.
2. Any found files will be parsed and customClass parameters stored in config.
3. All custom classes will be compiled and linked.
4. IOSTarget creates a new file `ib.classes` that contains a comma-delimited list with all custom classes.
5. At application launch, UIApplication.main reads that file and registers them as ObjCClass.

### CODE REVIEW

File: compiler/src/main/java/org/robovm/compiler/AppCompiler.java [hunk 1]
```diff
@@ -350,6 +372,63 @@ public void compile() throws IOException {
...
+                        throws IOException {
+                    
+                    String filename = file.getName().toLowerCase();
+                    if (filename.endsWith(".storyboard") || filename.endsWith(".xib")) {
+                        try {
+                            List<String> customClasses = findCustomClassesInIBFile(file);
+                            config.addCustomIBClasses(customClasses);
+                        } catch (XMLStreamException | IOException e) {
+                            // Storyboard or Xib may be corrupt.
+                            e.printStackTrace();
```
Comment (badlogic) [inline compiler/src/main/java/org/robovm/compiler/AppCompiler.java]: You probably want to use `config.getLogger().xxx` here.

File: compiler/src/main/java/org/robovm/compiler/AppCompiler.java [hunk 2]
```diff
@@ -196,6 +203,19 @@ public Config getConfig() {
             classes.add(clazz);
         }
         
+        if (!config.getCustomIBClasses().isEmpty()) {
+            for (String s : config.getCustomIBClasses()) {
+                if (s == null || s.trim().isEmpty()) {
```
Comment (badlogic) [inline compiler/src/main/java/org/robovm/compiler/AppCompiler.java]: You probably want to ensure that any custom class names added to `Config` are not null not empty before adding them to the `Config`. Config is kinda supposed to be immutable (it's not really anymore). But at least we should try not to return nulls or empty strings when not necessary. Do the check in line 390 and only add if the class name is != null and !empty.

File: compiler/src/main/java/org/robovm/compiler/AppCompiler.java [hunk 3]
```diff
@@ -350,6 +369,65 @@ public void compile() throws IOException {
...
+                    }
+                }
+            }, destDir);
+        }
+    }
+    
+    private List<String> findCustomClassesInIBFile(File file) throws XMLStreamException, IOException {
+        List<String> customClasses = new ArrayList<>();
+
+        FileInputStream fis = FileUtils.openInputStream(file);
```
Comment (ntherning) [inline compiler/src/main/java/org/robovm/compiler/AppCompiler.java]: Please use a try-with-resources block here

File: compiler/src/main/java/org/robovm/compiler/target/ios/IOSTarget.java [hunk 4]
```diff
@@ -443,6 +450,16 @@ private void strip(File dir, String executable) throws IOException {
                 .args("strip", "-x", new File(dir, executable))
                 .exec();
     }
+    
+    private void generateIBClassesFile(File installDir) throws IOException {
+        List<String> customClasses = config.getCustomIBClasses();
+        if (!customClasses.isEmpty()) {
+            FileWriter fw = new FileWriter(new File(installDir, "ib.classes"));
```
Comment (ntherning) [inline compiler/src/main/java/org/robovm/compiler/target/ios/IOSTarget.java]: Again, use try-with-resources.

File: cocoatouch/src/main/java/org/robovm/apple/uikit/UIApplication.java [hunk 5]
```diff
@@ -359,9 +360,37 @@ public void invoke(NSNotification a) {
...
     }
     
+    /**
+     * Registers all custom interface builder classes from the compiler generated ib.classes file, if any.
+     */
+    private static void registerCustomIBClasses() {
+        File ibClassesFile = new File(NSBundle.getMainBundle().getBundlePath() + "/ib.classes");
+        if (ibClassesFile.exists()) {
+            try {
+                BufferedReader reader = new BufferedReader(new FileReader(ibClassesFile));
```
Comment (ntherning) [inline cocoatouch/src/main/java/org/robovm/apple/uikit/UIApplication.java]: Sorry, missed this one the first time I had a look. try-with-resources again. :-)

**Comment by ntherning** [comment]
Before we merge this we have to test IPA creation for multiple archs and whether the IB classes in the config for each slice are propagated properly to the initial config. I suspect that they won't. We had to do a [hack](https://github.com/robovm/robovm/blob/master/compiler/src/main/java/org/robovm/compiler/AppCompiler.java#L644) in `AppCompiler.createIpa()` to copy over resource paths from the configs for each slice to the initial config. We probably have to do the same here.

**Comment by ntherning** [comment]
Looks good!

**Pull request merged**